replit_river/codegen/client.py

```diff
@@ -242,166 +246,222 @@ def generate_river_client_module(
                     chunks.extend(errors_chunks)
             else:
                 error_type = "RiverError"
-                output_or_error_type = f"Union[{output_type}, {error_type}]"
-                error_encoder = f"TypeAdapter({error_type}).validate_python(x)"
+            output_or_error_type = f"Union[{output_type}, {error_type}]"
+
+            # NB: These strings must be indented to at least the same level of
+            #     the function strings in the branches below, otherwise `dedent`
+            #     will pick our indentation level for normalization, which will
+            #     break the "def" indentation presuppositions.
+            parse_output_method = f"""\
+                                lambda x: TypeAdapter({output_type})
+                                    .validate_python(
+                                        x # type: ignore[arg-type]
+                                    )
+                """.strip()
+            parse_error_method = f"""\
+                                lambda x: TypeAdapter({error_type})
+                                    .validate_python(
+                                        x # type: ignore[arg-type]
+                                    )
+                """.strip()
 
-            output_encoder = f"TypeAdapter({output_type}).validate_python(x)"
             if output_type == "None":
-                output_encoder = "None"
-            # TODO: mypy ignore is added because TypeAdapter(...).validate_python
-            # cannot handle Union types, it should be fixed by making
-            # parse_output_method type aware.
-            parse_output_method = (
-                f"lambda x: {output_encoder}, # type: ignore[arg-type]"
-            )
-            parse_error_method = f"lambda x: {error_encoder}, # type: ignore[arg-type]"
+                parse_output_method = "lambda x: None"
 
             if procedure.type == "rpc":
                 control_flow_keyword = "return "
                 if output_type == "None":
                     control_flow_keyword = ""
                 current_chunks.extend(
                     [
-                        f"  async def {name}(",
-                        "    self,",
-                        f"    input: {input_type},",
-                        f"  ) -> {output_type}:"
-                        f"    {control_flow_keyword}await self.client.send_rpc(",
-                        f"      '{schema_name}',",
-                        f"      '{name}',",
-                        "      input,",
-                        f"      lambda x: TypeAdapter({input_type}).dump_python(x, ",
-                        "        by_alias=True,",
-                        "        exclude_none=True,",
-                        "      ),",
-                        f"     {parse_output_method}",
-                        f"     {parse_error_method}",
-                        "    )",
+                        indent(
+                            dedent(
```

Comment:
might be worth a comment, i would've been confused if not for the pr description